### PR TITLE
Initial Beacon P2P Protobufs, Subscriptions

### PR DESCRIPTION
**Pull request by rauljordan**
Hi all,

This PR uses the p2p shared service to create a beaconp2p workflow with subscriptions and a network simulator that will be part of the sync process.

### Requirements for Merge

- [x] Finish syncer workflow for receiving incoming blocks via feeds of proto messages
- [x] Finish tests for receiving incoming blocks appropriately

### CODE REVIEW

File: proto/sharding/v1/messages.proto [hunk 1]
```diff
@@ -7,8 +7,39 @@ enum Topic {
...
+  bytes parent_hash = 1;
+  uint64 slot_number = 2;
+  bytes randao_reveal = 3;
+  bytes attestation_bitmask = 4;
+  repeated uint64 attestation_aggregate_sig = 5;
+  repeated AggregateVote shard_aggregate_votes = 6;
+  bytes main_chain_ref = 7;
+  bytes active_state_hash = 8;
+  bytes crystallized_state_hash = 9;
+  string timestamp = 10;
```
Comment (prestonvanloon) [inline proto/sharding/v1/messages.proto]: Why string? This should be epoch time or some type import from well know types.

File: shared/p2p/service.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ import (
 	pb "github.com/prysmaticlabs/prysm/proto/sharding/v1"
 )
 
-// Sender represents a struct that is able to relay information via shardp2p.
+// Sender represents a struct that is able to relay information via p2o.
```
Comment (prestonvanloon) [inline shared/p2p/service.go]: Typo

File: beacon-chain/sync/service.go [hunk 3]
```diff
@@ -42,93 +49,96 @@ func DefaultConfig() Config {
...
+	ss.blockSub.Unsubscribe()
 	return nil
 }
 
 // ReceiveBlockHash accepts a block hash.
 // New hashes are forwarded to other peers in the network (unimplemented), and
 // the contents of the block are requested if the local chain doesn't have the block.
-func (ss *Service) ReceiveBlockHash(h hash.Hash) {
+func (ss *Service) ReceiveBlockHash(data *pb.BeaconBlockHashAnnounce) error {
+	h, err := blake2s.New256(data.Hash)
```
Comment (terenc3t) [inline beacon-chain/sync/service.go]: any reason we are using `blake2s` not `blake2b`?

File: beacon-chain/sync/service_test.go [hunk 4]
```diff
@@ -3,229 +3,198 @@ package sync
...
-	// both blocks are processed
-	testutil.AssertLogsContain(t, hook, fmt.Sprintf("broadcasting block: %x", h1.Sum(nil)))
-	testutil.AssertLogsContain(t, hook, fmt.Sprintf("forwarding block: %x", h1.Sum(nil)))
-	testutil.AssertLogsContain(t, hook, fmt.Sprintf("broadcasting block: %x", h2.Sum(nil)))
-	testutil.AssertLogsContain(t, hook, fmt.Sprintf("forwarding block: %x", h2.Sum(nil)))
+	// Sync service broadcasts the two separate blocks
+	// and forwards them to to the local chain.
+	testutil.AssertLogsContain(t, hook, "Broadcasting block hash to peers")
+	testutil.AssertLogsContain(t, hook, "Processed block")
+	testutil.AssertLogsContain(t, hook, "Broadcasting block hash to peers")
```
Comment (rawfalafel) [inline beacon-chain/sync/service_test.go]: This doesn't test that two different blocks were processed. A potential fix is to include the hash in `BeaconBlockResponse, log the hash, and update this assertion.

File: WORKSPACE [hunk 5]
```diff
@@ -10,8 +10,7 @@ http_archive(
```
Comment (prestonvanloon) [inline WORKSPACE]: Please add comment here saying we cannot update until https://github.com/bazelbuild/bazel-gazelle/issues/272 is resolved.

File: beacon-chain/node/node.go [hunk 6]
```diff
@@ -11,13 +11,13 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/prysmaticlabs/prysm/beacon-chain/blockchain"
 	"github.com/prysmaticlabs/prysm/beacon-chain/database"
-	"github.com/prysmaticlabs/prysm/beacon-chain/network"
 	"github.com/prysmaticlabs/prysm/beacon-chain/powchain"
 	rbcSync "github.com/prysmaticlabs/prysm/beacon-chain/sync"
```
Comment (terenc3t) [inline beacon-chain/node/node.go]: s/rbcSync/rbcsync? 

I think the style guide doesn't recommend mixed case for package name
https://blog.golang.org/package-names

File: beacon-chain/sync/service.go [hunk 7]
```diff
@@ -23,12 +26,12 @@ var log = logrus.WithField("prefix", "sync")
...
-	blockBuf       chan *types.Block
+	ctx                  context.Context
+	cancel               context.CancelFunc
+	p2p                  types.P2P
+	chainService         types.ChainService
+	announceBlockHashBuf chan p2p.Message
+	blockBuf             chan p2p.Message
 }
 
 // Config allows the channel's buffer sizes to be changed
```
Comment (terenc3t) [inline beacon-chain/sync/service.go]: punctuation

File: beacon-chain/types/block.go [hunk 8]
```diff
@@ -1,72 +1,120 @@
...
+	"github.com/golang/protobuf/ptypes"
+	pb "github.com/prysmaticlabs/prysm/proto/sharding/v1"
 	"golang.org/x/crypto/blake2b"
 )
 
 // Block defines a beacon chain core primitive.
 type Block struct {
-	data *Data
+	data                  *pb.BeaconBlockResponse
+	activeStateHash       hash.Hash
```
Comment (terenc3t) [inline beacon-chain/types/block.go]: why aren't `activeStateHash` and `crystallizedStateHash` part of `data`?

File: beacon-chain/types/block.go [hunk 9]
```diff
@@ -1,72 +1,120 @@
...
-	SlotNumber              uint64          // Slot number is the number a client should check to know when it creates block.
-	RandaoReveal            hash.Hash       // RandaoReveal is used for Randao commitment reveal.
-	AttestationBitmask      []byte          // AttestationBitmask is the bit field of who from the attestation committee participated.
-	AttestationAggregateSig []uint          // AttestationAggregateSig is validator's aggregate sig.
-	ShardAggregateVotes     []AggregateVote // ShardAggregateVotes is shard aggregate votes.
-	MainChainRef            common.Hash     // MainChainRef is the reference to main chain block.
-	ActiveStateHash         hash.Hash       // ActiveStateHash is the state that changes every block.
-	CrystallizedStateHash   hash.Hash       // CrystallizedStateHash is the state that changes every epoch.
-	Timestamp               time.Time
+// RandaoReveal returns the blake2b randao hash.
```
Comment (terenc3t) [inline beacon-chain/types/block.go]: Can you please add a comment this is just temporary?  Randao reveal happens after proposer signs a block which generates a random number

File: beacon-chain/sync/service.go [hunk 10]
```diff
@@ -92,43 +74,71 @@ func (ss *Service) Stop() error {
...
 			return
-		case h := <-hashBuf:
-			ps.RequestBlock(h)
-		case b := <-blockBuf:
-			cs.ProcessBlock(b)
+		case msg := <-ss.announceBlockHashBuf:
+			data, ok := msg.Data.(pb.BeaconBlockHashAnnounce)
+			// TODO: Handle this at p2p layer.
+			if !ok {
+				log.Errorf("Received malformed beacon block hash announcement p2p message")
```
Comment (nisdas) [inline beacon-chain/sync/service.go]: change this to `log.Error`

File: beacon-chain/types/block.go (hunk 8 above)
Comment (nisdas) [reply beacon-chain/types/block.go]: I think they are, `activeStateHash ` and `crystallizedStateHash` does not need to be added to the block struct over here, we can get them from `data`

**Pull request merged**